### PR TITLE
Add support for passing files and markdown directly to `Show-Markdown`

### LINKED ISSUE #7338: Showing markdown with new commands is harder than it should be
When I first tried to use the new markdown commands to show a markdown file, it seemed obvious to me that I'd just run:

```
# Expected:
Show-Markdown .\README.md
```
but that doesn't work.  You have to execute this instead:
```
# Actual:
ConvertFrom-Markdown -path .\README.md -AsVT100EncodedString | Show-Markdown
```

It seems to me that `Show-Markdown` should have a Path parameter set so that you can simply specify the path to a markdown file to be shown.

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      6.1.0-preview.4
PSEdition                      Core
GitCommitId                    6.1.0-preview.4
OS                             Microsoft Windows 10.0.17134
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
## PR Summary

Previously you had to use `ConvertFrom-Markdown` and pipe that output to `Show-Markdown` making it cumbersome to use interactively.  Added `-Path` and `-LiteralPath` parameters as well as allowing markdown text to passed directly as `-InputObject` (or via pipeline) to `Show-Markdown`.

Internally, `Show-Markdown` uses `ConvertFrom-Markdown` to handle these new capabilities.  Removed the use of `Out-Default` which only passed the resulting vt100 text to the console and instead using WriteObject() so that the output can be redirected or captured by scripts or further in the pipeline.  The testhook is still needed when `-UseBrowser` is specified.

Note I had to fully qualify `PowerShell` and `Path` types due to conflict with `Path` parameter and `PowerShell` namespace.

Fix https://github.com/PowerShell/PowerShell/issues/7338

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs [hunk 1]
```diff
@@ -28,46 +30,119 @@ public class ShowMarkdownCommand : PSCmdlet
...
+            if (inputObject is MarkdownInfo markdownInfo)
             {
                 if (UseBrowser)
                 {
                     var html = markdownInfo.Html;
 
                     if (!string.IsNullOrEmpty(html))
                     {
-                        string tmpFilePath = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString() + ".html");
+                        string tmpFilePath = System.IO.Path.Combine(System.IO.Path.GetTempPath(), Guid.NewGuid().ToString() + ".html");
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs]: I believe we can remove `System.IO`.

**Review by iSazonov — APPROVED** [review]
LGTM with one minor comment.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs (hunk 1 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs]: The compiler complained that `Path` type here was confused with the `Path` parameter.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs [hunk 2]
```diff
@@ -28,46 +30,119 @@ public class ShowMarkdownCommand : PSCmdlet
...
         /// </summary>
         protected override void BeginProcessing()
         {
-            if (!UseBrowser.IsPresent)
-            {
-                // Since UseBrowser is not bound, we use proxy to Out-Default
-                stepPipe = ScriptBlock.Create(@"Microsoft.PowerShell.Core\Out-Default @PSBoundParameters").GetSteppablePipeline(this.MyInvocation.CommandOrigin);
-                stepPipe.Begin(this);
-            }
+            _powerShell = System.Management.Automation.PowerShell.Create();
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs]: This constructor will create a new Runspace. Please use `PowerShell.Create(RunspaceMode.CurrentRunspace)` instead, which will use a nested pipeline in the same Runsapce.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs [hunk 3]
```diff
@@ -28,46 +30,119 @@ public class ShowMarkdownCommand : PSCmdlet
...
         /// <summary>
         /// Override ProcessRecord.
         /// </summary>
         protected override void ProcessRecord()
         {
-            object inpObj = InputObject.BaseObject;
+            switch (ParameterSetName)
+            {
+                case "InputObject":
+                    if (InputObject.BaseObject is MarkdownInfo)
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs]: Use `is MarkdownInfo markdownInfo`, and pass in `markdownInfo` instead.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs [hunk 4]
```diff
@@ -28,46 +30,119 @@ public class ShowMarkdownCommand : PSCmdlet
...
+                    break;
+
+                case "Path":
+                case "LiteralPath":
+                    ConvertFromMarkdown(ParameterSetName, Path);
+                    break;
+
+                default:
+                    Debug.Assert(false, string.Format(CultureInfo.InvariantCulture, "Invalid parameter set name: {0}", ParameterSetName));
+                    break;
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs]: Please make it throw an `InvalidOperationException`. The assertion is not that useful with our current building/testing.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs [hunk 5]
```diff
@@ -28,46 +30,119 @@ public class ShowMarkdownCommand : PSCmdlet
...
+        /// <param name="input">Value of parameter.</param>
+        private void ConvertFromMarkdown(string parameter, object input)
+        {
+            _powerShell.AddCommand("Microsoft.PowerShell.Utility\\ConvertFrom-Markdown").AddParameter(parameter, input);
+            if (!UseBrowser)
+            {
+                _powerShell.AddParameter("AsVT100EncodedString");
+            }
+
+            Collection<PSObject> output = _powerShell.Invoke();
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs]: You can use `Invoke<MarkdownInfo>()`, so the output will be `MarkdownInfo` collection and we can avoid the type comparison in `ProcessMarkdownInfo`.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs [hunk 6]
```diff
@@ -28,46 +30,119 @@ public class ShowMarkdownCommand : PSCmdlet
...
+        }
 
-            if (inpObj is MarkdownInfo markdownInfo)
+        /// <summary>
+        /// Process markdown as input objects.
+        /// </summary>
+        /// <param name="inputObject">Markdown object to process.</param>
+        private void ProcessMarkdownInfo(object inputObject)
+        {
+            if (inputObject is MarkdownInfo markdownInfo)
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs]: If you address the comments above, this check will no longer be needed.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs [hunk 7]
```diff
@@ -157,12 +223,12 @@ protected override void ProcessRecord()
...
             else
             {
-                string errorMessage = StringUtil.Format(ConvertMarkdownStrings.InvalidInputObjectType, inpObj.GetType());
+                string errorMessage = StringUtil.Format(ConvertMarkdownStrings.InvalidInputObjectType, inputObject.GetType());
                 var errorRecord = new ErrorRecord(
                             new ArgumentException(errorMessage),
                             "InvalidInputObject",
                             ErrorCategory.InvalidArgument,
-                            InputObject);
+                            inputObject);
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs]: This error message won't be needed if you address the above comments.

**Review by daxian-dbw — CHANGES_REQUESTED** [review]
Two more comments:
1. `ConvertFrom-Markdown e:\powershell\readme.md` throws error: `A positional parameter cannot be found that accepts argument 'E:\PowerShell\README.md'`. I think it should work.

2. `ConvertFrom-Markdown -Path E:\PowerShell\README.md | Show-Markdown` fails with this error: `Show-Markdown : The property VT100EncodedString of the given object is null or empty.` So, `ConvertFrom-Markdown` by default only populates HTML, while `Show-Markdown` by default show VT100. That feels very inconsistent to me.

Opened issue #7363

**Comment by SteveL-MSFT** [comment]
@daxian-dbw I think the `ConvertFrom-Markdown` issues you bring up should be a separate issue

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs [hunk 8]
```diff
@@ -173,9 +239,10 @@ protected override void ProcessRecord()
         /// </summary>
         protected override void EndProcessing()
         {
-            if (stepPipe != null)
+            if (_powerShell != null)
             {
-                stepPipe.End();
+                _powerShell.Stop();
```
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs]: Nitpick.  You don't need to call Stop().  Dispose() will do this for you.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs [hunk 9]
```diff
@@ -28,143 +30,192 @@ public class ShowMarkdownCommand : PSCmdlet
...
+                        "HtmlIsNullOrEmpty",
+                        ErrorCategory.InvalidData,
+                        html);
+
+                    WriteError(errorRecord);
+                }
             }
             else
             {
-                string errorMessage = StringUtil.Format(ConvertMarkdownStrings.InvalidInputObjectType, inpObj.GetType());
```
Comment (daxian-dbw) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs]: > ConvertMarkdownStrings.InvalidInputObjectType

I guess this resource string can be removed now.

**Review by PaulHigin — APPROVED** [review]
LGTM

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs (hunk 9 above)
Comment (SteveL-MSFT) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/ShowMarkdownCommand.cs]: That string is still used by `ConvertFrom-Markdown` and `Set-MarkdownOption` cmdlets

**Review by daxian-dbw — APPROVED** [review]
LGTM

**Comment by SteveL-MSFT** [comment]
@dantraMSFT CI is failing on macOS oslog tests.  I don't see any thing related to the changes I've made.  Can you take a look?

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

**Review by adityapatwardhan — APPROVED** [review]

**Pull request merged**